File: apps/demo-be-python-fastapi/tests/unit/steps/unit_handling_steps.py
"""Unit BDD step definitions for unit handling feature."""

import json

import pytest
from fastapi.testclient import TestClient
from pytest_bdd import given, parsers, scenarios, then, when

from tests.unit.conftest import GHERKIN_ROOT

pytestmark = pytest.mark.unit

scenarios(str(GHERKIN_ROOT / "expenses" / "unit-handling.feature"))

_PASSWORD = "Str0ng#Pass1"


@given(
    '"alice" has logged in and stored the access token',
    target_fixture="alice_tokens",
)
def alice_login_units(client: TestClient, registered_user: dict) -> dict:
    resp = client.post(
        "/api/v1/auth/login",
        json={"username": "alice", "password": _PASSWORD},
    )
    assert resp.status_code == 200
    return resp.json()


@given(
    parsers.parse("alice has created an expense with body {body}"),
    target_fixture="created_expense",
)
def alice_create_unit_expense(client: TestClient, alice_tokens: dict, body: str) -> dict:
    data = json.loads(body)
    resp = client.post(
        "/api/v1/expenses",
        json=data,
        headers={"Authorization": f"Bearer {alice_tokens['access_token']}"},
    )
    assert resp.status_code == 201, f"Create expense failed: {resp.text}"
    return resp.json()


# --- When steps ---


@when("alice sends GET /api/v1/expenses/{expenseId}", target_fixture="response")
def alice_get_unit_expense(client: TestClient, alice_tokens: dict, created_expense: dict):  # type: ignore[no-untyped-def]
    return client.get(
        f"/api/v1/expenses/{created_expense['id']}",
        headers={"Authorization": f"Bearer {alice_tokens['access_token']}"},
    )


@when(
    parsers.parse("alice sends POST /api/v1/expenses with body {body}"),
    target_fixture="response",
)
def alice_post_unit_expense(client: TestClient, alice_tokens: dict, body: str):  # type: ignore[no-untyped-def]
    data = json.loads(body)
    return client.post(
        "/api/v1/expenses",
        json=data,
        headers={"Authorization": f"Bearer {alice_tokens['access_token']}"},
    )


# --- Then steps ---


@then(parsers.parse('the response body should contain "quantity" equal to {value}'))
def check_quantity(response, value: str) -> None:
    body = response.json()
    assert "quantity" in body, f"'quantity' not in response: {body}"
    actual = body["quantity"]
    expected = float(value)
    assert float(actual) == expected, f"Expected quantity={expected}, got {actual}"


@then(parsers.parse('the response body should contain "unit" equal to "{value}"'))
def check_unit(response, value: str) -> None:
    body = response.json()
    assert "unit" in body, f"'unit' not in response: {body}"
    assert body["unit"] == value, f"Expected unit={value!r}, got {body['unit']!r}"

File: apps/demo-fs-ts-nextjs/test/unit/be-steps/unit-handling.steps.ts
import path from "path";
import { loadFeature, describeFeature } from "@amiceli/vitest-cucumber";
import { expect } from "vitest";
import { createTestContext, registerUser, loginUser, getAuth, type TestContext } from "./helpers/test-context";

const feature = await loadFeature(
  path.resolve(process.cwd(), "../../specs/apps/demo/be/gherkin/expenses/unit-handling.feature"),
);

async function createExpense(ctx: TestContext, body: Record<string, unknown>): Promise<string> {
  const resp = await ctx.client.dispatch("POST", "/api/v1/expenses", body, getAuth(ctx, "alice"));
  if (resp.status !== 201) throw new Error(`Failed: ${JSON.stringify(resp.body)}`);
  return (resp.body as { id: string }).id;
}

describeFeature(feature, ({ Scenario, Background }) => {
  let ctx: TestContext;

  Background(({ Given, And }) => {
    Given("the API is running", () => {
      ctx = createTestContext();
    });

    And('a user "alice" is registered with email "alice@example.com" and password "Str0ng#Pass1"', async () => {
      await registerUser(ctx, "alice", "alice@example.com", "Str0ng#Pass1");
    });

    And('"alice" has logged in and stored the access token', async () => {
      await loginUser(ctx, "alice", "Str0ng#Pass1");
    });
  });

  Scenario(
    'Create expense with metric unit "liter" stores quantity and unit correctly',
    ({ Given, When, Then, And }) => {
      Given(
        'alice has created an expense with body { "amount": "75000", "currency": "IDR", "category": "fuel", "description": "Petrol", "date": "2025-01-15", "type": "expense", "quantity": 50.5, "unit": "liter" }',
        async () => {
          ctx.context.expenseId = await createExpense(ctx, {
            amount: "75000",
            currency: "IDR",
            category: "fuel",
            description: "Petrol",
            date: "2025-01-15",
            type: "expense",
            quantity: "50.5",
            unit: "liter",
          });
        },
      );

      When("alice sends GET /api/v1/expenses/{expenseId}", async () => {
        ctx.response = await ctx.client.dispatch(
          "GET",
          `/api/v1/expenses/${ctx.context.expenseId}`,
          null,
          getAuth(ctx, "alice"),
        );
      });

      Then("the response status code should be 200", () => {
        expect(ctx.response!.status).toBe(200);
      });

      And('the response body should contain "quantity" equal to 50.5', () => {
        expect(parseFloat(String((ctx.response!.body as Record<string, unknown>).quantity))).toBe(50.5);
      });

      And('the response body should contain "unit" equal to "liter"', () => {
        expect((ctx.response!.body as Record<string, unknown>).unit).toBe("liter");
      });
    },
  );

  Scenario(
    'Create expense with imperial unit "gallon" stores quantity and unit correctly',
    ({ Given, When, Then, And }) => {
      Given(
        'alice has created an expense with body { "amount": "45.00", "currency": "USD", "category": "fuel", "description": "Gas", "date": "2025-01-15", "type": "expense", "quantity": 10, "unit": "gallon" }',
        async () => {
          ctx.context.expenseId = await createExpense(ctx, {
            amount: "45.00",
            currency: "USD",
            category: "fuel",
            description: "Gas",
            date: "2025-01-15",
            type: "expense",
            quantity: "10",
            unit: "gallon",
          });
        },
      );

      When("alice sends GET /api/v1/expenses/{expenseId}", async () => {
        ctx.response = await ctx.client.dispatch(
          "GET",
          `/api/v1/expenses/${ctx.context.expenseId}`,
          null,
          getAuth(ctx, "alice"),
        );
      });

      Then("the response status code should be 200", () => {
        expect(ctx.response!.status).toBe(200);
      });

      And('the response body should contain "quantity" equal to 10', () => {
        expect(parseFloat(String((ctx.response!.body as Record<string, unknown>).quantity))).toBe(10);
      });

      And('the response body should contain "unit" equal to "gallon"', () => {
        expect((ctx.response!.body as Record<string, unknown>).unit).toBe("gallon");
      });
    },
  );

  Scenario("Create expense with an unsupported unit returns 400", ({ When, Then, And }) => {
    When(
      'alice sends POST /api/v1/expenses with body { "amount": "10.00", "currency": "USD", "category": "misc", "description": "Cargo", "date": "2025-01-15", "type": "expense", "quantity": 5, "unit": "fathom" }',
      async () => {
        ctx.response = await ctx.client.dispatch(
          "POST",
          "/api/v1/expenses",
          {
            amount: "10.00",
            currency: "USD",
            category: "misc",
            description: "Cargo",
            date: "2025-01-15",
            type: "expense",
            quantity: "5",
            unit: "fathom",
          },
          getAuth(ctx, "alice"),
        );
      },
    );

    Then("the response status code should be 400", () => {
      expect(ctx.response!.status).toBe(400);
    });

    And('the response body should contain a validation error for "unit"', () => {
      const body = ctx.response!.body as Record<string, unknown>;
      expect(String(body.error).toLowerCase()).toContain("unit");
    });
  });

  Scenario("Expense without quantity and unit fields is accepted", ({ When, Then, And }) => {
    When(
      'alice sends POST /api/v1/expenses with body { "amount": "25.00", "currency": "USD", "category": "food", "description": "Dinner", "date": "2025-01-15", "type": "expense" }',
      async () => {
        ctx.response = await ctx.client.dispatch(
          "POST",
          "/api/v1/expenses",
          {
            amount: "25.00",
            currency: "USD",
            category: "food",
            description: "Dinner",
            date: "2025-01-15",
            type: "expense",
          },
          getAuth(ctx, "alice"),
        );
      },
    );

    Then("the response status code should be 201", () => {
      expect(ctx.response!.status).toBe(201);
    });

    And('the response body should contain a non-null "id" field', () => {
      expect((ctx.response!.body as Record<string, unknown>).id).toBeDefined();
    });
  });
});
